### DISCUSSION TITLE
[Application Development] Trace the transactions from L2 to L1 || How to get confirmations on L1

### DISCUSSION THREAD

**Discussion by threetoeddinosaur** [ethereum-optimism/developers#627] [Application Development]
### Did you check the documentation?

- [X] Yes
- [ ] No

### What type of issue are you experiencing?

- [ ] Error
- [ ] Feature Request
- [X] General Question

### Issue Description

I'm trying to see how I can get from an L2 transaction batch to a L1 tx confirming the inclusion of the block containing the L2 transaction into a batch which was submitted to L1. If I understand correctly, batches are constructed from a range of L2 blocks and the stateRoot of each block goes into the batch. That batch is then stored on the L1 as a blob where the transaction was made by the Batch Sender address. I can see the blobs sent by the batch sender, but I can find the link from a particular block on L2 to a particular batch, and/or the specific batch to the blob commitment relating to that batch? I need to be able to provably tie a specific L2 transaction to a specific piece of L1 data showing it happened. I am aware of the getBlockByNumber RPC call, but that is not sufficient, it only tells me that my L2 transaction is finalized on L1, and doesn't provide any information I can look at on L1 to prove that.

### Steps to Reproduce

Not a bug, general question

### Are you using a specific library or SDK?

- [ ] OP-SDK
- [ ] Viem
- [ ] Other
- [X] Not Sure

### Are you using the latest version of the library/SDK?

- [ ] Yes
- [X] No

### What type of support do you need?

- [X] Technical support for issue resolution
- [ ] Feature implementation guidance
- [ ] General advice or consultation
- [X] Documentation or learning resources

**Comment by hash1go** [comment]
Hello!
Firstly, to clarify, the blob does not contain the L2 state root information. Please refer to this specification for further details.
https://specs.optimism.io/protocol/derivation.html#sequencing--batch-submission-overview

One key point to note is that L2 blocks are derived from L1 data, so the concept of directly linking an L2 transaction to L1 data can be misleading. Instead, to confirm that the transaction execution on L2 aligns with the commitments submitted to L1, you can check the output proposal submitted by the op-proposer. Since the OP Stack functions as an Optimistic Rollup, this output proposal is considered valid unless challenged, confirming that L2 blocks have progressed correctly based on L1.

**Reply by deqcjain** [reply]
Hi @hash1go, How can we get the confirmations number on L1 for a transaction that happened on L2? Also how can we check the proposal as there was one way I know is through the L2OutputOracle contract (deprecated)

**Reply by hash1go** [reply]
For sync status, this document should be helpful.
https://docs.optimism.io/builders/node-operators/json-rpc#optimism_syncstatus
https://docs.optimism.io/connect/resources/glossary#safe-l2-block

Also, to verify output proposals, you can either run your own L1 execution client or trust and use a third-party RPC provider.

This kind of explorer is also useful.
https://optimism.blockscout.com/dispute-games

**Reply by threetoeddinosaur** [reply]
@hash1go thanks for responding, much appreciated
So, knowing the sequencer constructs the block as well as batches, and both the block and the batch can be linked through block number or timestamp (maybe).  Multiple batches are encoded into channel and (1channel = 1 L1epoch). But what is the way I can get that link as well the L1 transaction where this batch has been submitted.

**Reply by hash1go** [reply]
There's no direct way to track which L1 data contains a specific L2 transaction. You can either check `optimism_syncStatus` or manually decode the L1 blob data.

**Reply by deqcjain** [reply]
> There's no direct way to track which L1 data contains a specific L2 transaction. You can either check `optimism_syncStatus` or manually decode the L1 blob data.

But which L1 transaction contains the blob for my L2 block

**Reply by hash1go** [reply]
Do you want to know the op-batcher address? You can check it here.
https://blobscan.com/address/0x6887246668a3b87f54deb3b94ba47a6f63f32985

**Reply by joohhnnn** [reply]
> @hash1go thanks for responding, much appreciated So, knowing the sequencer constructs the block as well as batches, and both the block and the batch can be linked through block number or timestamp (maybe). Multiple batches are encoded into channel and (1channel = 1 L1epoch). But what is the way I can get that link as well the L1 transaction where this batch has been submitted.

You can filter out the L1 blocks that contain batch transactions from the batcher, and then use `optimism_safeHeadAtL1Block` to determine which L2 block/transactions are corresponding to the specific L1 block.

**Reply by threetoeddinosaur** [reply]
@joohhnnn this RPC call as I know takes an L1 blockNumber(hexadecimal) and gives output that in which L1 block the L2 Block is. Am I correct? 
So, I made this call below, I got this L1 Block from the Batcher tx for this L2 Block 17822386
```
{
    "jsonrpc": "2.0",
    "method": "optimism_safeHeadAtL1Block",
    "params": [
        "0x6bc075"//7061621
    ],
    "id": 1
}
``` 
and got this response
`{
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
        "l1Block": {
            "hash": "0xb9a609a8fe503dda37a96d3a8886f3e7cdeedbd986e60195fcb0cdbab4346558",
            "number": 7061600
        },
        "safeHead": {
            "hash": "0x3132bbfd11ad37213aa7f12d614e51a9549a87534669474769cafedb074ccd65",
            "number": 17822075
        }
    }
}`

I am not able to understand the response as the L1 Block for which I am making the call is ahead of the L1Block in response and the L2BlockNumber is way ahead for which I got the L1 Block. Can you please explain why is that.

**Reply by chahatjaink** [reply]
Same issue

**Reply by joohhnnn** [reply]
> I am not able to understand the response as the L1 Block for which I am making the call is ahead of the L1Block in response and the L2BlockNumber is way ahead for which I got the L1 Block. Can you please explain why is that.

PLZ check the whole conversation in the discord. That may give a help @threetoeddinosaur @chahatjaink 
https://discord.com/channels/667044843901681675/1080862392281481246/1301445127075659798

**Reply by chahatjaink** [reply]
@joohhnnn, In that conversation, you mentioned using optimism_safeHeadAtL1Block to identify which L2 blocks or transactions correspond to a specific L1 block. The output I'm seeing indicates that, when calling the safeHeadAtL1Block RPC method with L1BlockNumber = 7061621, the response returned is L1Block = 7061600 and L2SafeHead = 17822075.

I'm getting this L1BlockNumber input from the setL1BlockValues function call within each L2 block(17822386).  Could you please clarify what this input—specifically the L1BlockNumber—is intended to specify and why am I getting this response for the L1 Block that is way behind this one?

**Reply by threetoeddinosaur** [reply]
Hey @joohhnnn, So, I will explain what the issue I am facing is that I can get the L1Block for the L2 Block with the help of the setL1BlockValue method call in each L2 Block as the 0th index tx. For this L1Block I am calling the given optimism_safeHeadAtL1Block and getting back a response like the one described above, and safeHead for L2 Block is also less than my L2 Block (does this mean that my L2Block has not mined on L1 yet).

**Reply by joohhnnn** [reply]
> @joohhnnn, In that conversation, you mentioned using optimism_safeHeadAtL1Block to identify which L2 blocks or transactions correspond to a specific L1 block. The output I'm seeing indicates that, when calling the safeHeadAtL1Block RPC method with L1BlockNumber = 7061621, the response returned is L1Block = 7061600 and L2SafeHead = 17822075.
> 
> I'm getting this L1BlockNumber input from the setL1BlockValues function call within each L2 block(17822386). Could you please clarify what this input—specifically the L1BlockNumber—is intended to specify and why am I getting this response for the L1 Block that is way behind this one?

Hello, the 'input L1BlockNumber' represents the L1 block of the most recent batch transaction sent by the batcher up to and including the current or previous input. Therefore, when you enter 7061621, it will return 7061600, which means that the most recent batch transaction before block 7061621 occurred at block 7061600 and corresponds to L2 block 17822075.

@chahatjaink

**Reply by joohhnnn** [reply]
> Hey @joohhnnn, So, I will explain what the issue I am facing is that I can get the L1Block for the L2 Block with the help of the setL1BlockValue method call in each L2 Block as the 0th index tx. For this L1Block I am calling the given optimism_safeHeadAtL1Block and getting back a response like the one described above, and safeHead for L2 Block is also less than my L2 Block (does this mean that my L2Block has not mined on L1 yet).

When you use 'setL1BlockValue' to set the L1 block for L2, it does not mean that the L2 data is included in L1. Only when the batcher batches your L2 data to L1 does it correspond accordingly @threetoeddinosaur
